plugins/net_plugin/net_plugin.cpp

```diff
@@ -1781,14 +1813,30 @@ namespace eosio {
    }
 
    // called from connection strand
-   void connection::enqueue_block( const signed_block_ptr& b, bool to_sync_queue) {
+   size_t connection::enqueue_block( const signed_block_ptr& b, bool to_sync_queue) {
       peer_dlog( this, "enqueue block ${num}", ("num", b->block_num()) );
       verify_strand_in_this_thread( strand, __func__, __LINE__ );
 
       block_buffer_factory buff_factory;
       auto sb = buff_factory.get_send_buffer( b );
       latest_blk_time = std::chrono::system_clock::now();
-      enqueue_buffer( sb, no_reason, to_sync_queue);
+      if( block_sync_rate_limit > 0 ) {
+         int sleep_time_us = 100;
+         const int max_sleep_time_us = 100000;
+         while( true) {
+            auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(get_time() - connection_start_time);
+            auto current_rate = double(block_sync_bytes_sent) / elapsed.count();
+            if( current_rate < block_sync_rate_limit ) {
+               enqueue_buffer( sb, no_reason, to_sync_queue);
+               break;
+            }
+            usleep(sleep_time_us);
```

Comment:
Instead of sleeping can you just return and wait for the next call to `enqueue_block`.

Reply:
Simply returning doesn't work.  Blocks are lost and the receiving node starts getting nothing but unlinkable blocks, after a gap of almost 50 blocks.  At this level in the call stack, it is assumed the method never fails.

Reply:
This code needs to be moved to `enqueue_sync_block` and then don't update `peer_requested->last` and don't pull the block out of the controller when you throttle.

Reply:
Moved.

(To where it was intended to go in the first place, or why change the function signature... :expressionless:)